File: crates/nucleation-routing/src/drc.rs
//! Design-rule checking: shorts, support, decay, and directed cycles
//! through repeaters.
//!
//! The cycle check is the NEW tool the rca_cells session was missing: two
//! cells' port approaches formed opposite-facing repeaters in a ring on an
//! aliased net — no short, passed every static check, settled quiescent,
//! and latched the placement transient. A directed cycle through repeaters
//! on a net is a storage element; unless you built a latch on purpose, it
//! is a bug.

use crate::audit;
use crate::blocks::{self, is_comparator, is_dust, is_repeater};
use crate::nets;
use crate::workspace::Workspace;
use pnr_core::unionfind::UnionFind;
use pnr_core::Pos;
use std::collections::{BTreeMap, BTreeSet};

/// A DRC violation.
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Violation {
    /// Two distinct labels share an electrical net.
    Short {
        /// First label.
        label_a: String,
        /// Second label.
        label_b: String,
        /// Witness cell for the first label.
        at_a: Pos,
        /// Witness cell for the second label.
        at_b: Pos,
    },
    /// A block needing a floor has none.
    Floating {
        /// Where.
        at: Pos,
        /// Block name (without properties).
        block: String,
    },
    /// A wall torch's anchor block is missing or not solid.
    UnattachedWallTorch {
        /// The torch.
        at: Pos,
        /// The missing anchor cell.
        anchor: Pos,
    },
    /// A directed cycle through repeaters/comparators — an accidental
    /// latch. Lists the diodes on the cycle.
    RepeaterCycle {
        /// Repeater/comparator positions forming the ring.
        diodes: Vec<Pos>,
    },
    /// Dust farther from every driver than a full signal reaches.
    PowerStarved {
        /// The starved dust cell.
        at: Pos,
        /// Its distance from the nearest driver.
        distance: u32,
    },
}

/// DRC options.
#[derive(Clone, Debug, Default)]
pub struct DrcOptions {
    /// Label pairs that are deliberately the same electrical net.
    pub aliases: Vec<(String, String)>,
    /// Skip the decay (PowerStarved) check.
    pub skip_decay: bool,
}

/// Run all checks over a workspace.
pub fn drc(ws: &Workspace, opts: &DrcOptions) -> Vec<Violation> {
    let mut out = Vec::new();
    for s in nets::check(ws.cells(), ws.labels(), &opts.aliases) {
        out.push(Violation::Short {
            label_a: s.label_a,
            label_b: s.label_b,
            at_a: s.at_a,
            at_b: s.at_b,
        });
    }
    let a = audit::audit(ws.cells());
    for (at, block, _below) in a.floating {
        out.push(Violation::Floating { at, block });
    }
    for (at, _face, anchor) in a.unattached_wall_torch {
        out.push(Violation::UnattachedWallTorch { at, anchor });
    }
    out.extend(repeater_cycles(ws.cells()));
    if !opts.skip_decay {
        out.extend(decay_check(ws.cells(), 15));
    }
    out
}

/// The input and output cells of a repeater/comparator: `facing` names the
/// INPUT side (verified: `repeater[facing=west]` conducts toward +X).
fn diode_io(p: Pos, block: &str) -> Option<(Pos, Pos)> {
    let f = blocks::facing_of(block)?;
    let (dx, dy, dz) = blocks::facing_vec(f)?;
    Some((p.offset(dx, dy, dz), p.offset(-dx, -dy, -dz)))
}

/// Detect directed cycles through repeaters/comparators over the dust-net
/// graph: dust components are nodes, each diode is a directed edge from the
/// component at its input to the component at its output.
pub fn repeater_cycles(cells: &BTreeMap<Pos, String>) -> Vec<Violation> {
    // Dust components.
    let mut uf: UnionFind<Pos> = UnionFind::new();
    let dust: Vec<Pos> = cells
        .iter()
        .filter(|(_, b)| is_dust(b))
        .map(|(p, _)| *p)
        .collect();
    for p in &dust {
        uf.find(p);
        for q in nets::neighbours(cells, *p) {
            uf.union(p, &q);
        }
    }
    // Directed edges: component -> component, tagged with the diode.
    let mut edges: BTreeMap<Pos, Vec<(Pos, Pos)>> = BTreeMap::new(); // from-comp -> [(to-comp, diode)]
    for (p, b) in cells {
        if !(is_repeater(b) || is_comparator(b)) {
            continue;
        }
        let Some((inp, outp)) = diode_io(*p, b) else {
            continue;
        };
        let in_dust = cells.get(&inp).is_some_and(|b| is_dust(b));
        let out_dust = cells.get(&outp).is_some_and(|b| is_dust(b));
        if in_dust && out_dust {
            let a = uf.find(&inp);
            let bcomp = uf.find(&outp);
            edges.entry(a).or_default().push((bcomp, *p));
        }
    }
    // DFS cycle detection over component nodes (deterministic order).
    #[derive(Clone, Copy, PartialEq)]
    enum Color {
        White,
        Grey,
        Black,
    }
    let nodes: BTreeSet<Pos> = edges
        .iter()
        .flat_map(|(from, tos)| std::iter::once(*from).chain(tos.iter().map(|(t, _)| *t)))
        .collect();
    let mut color: BTreeMap<Pos, Color> = nodes.iter().map(|n| (*n, Color::White)).collect();
    let mut found: Vec<Violation> = Vec::new();
    let mut reported: BTreeSet<Vec<Pos>> = BTreeSet::new();

    fn dfs(
        node: Pos,
        edges: &BTreeMap<Pos, Vec<(Pos, Pos)>>,
        color: &mut BTreeMap<Pos, Color>,
        stack: &mut Vec<(Pos, Pos)>, // (component, diode entered through)
        found: &mut Vec<Violation>,
        reported: &mut BTreeSet<Vec<Pos>>,
    ) {
        color.insert(node, Color::Grey);
        if let Some(succs) = edges.get(&node) {
            for (to, diode) in succs {
                match color.get(to).copied().unwrap_or(Color::White) {
                    Color::Grey => {
                        // Cycle: the diodes are the edges taken from `to`
                        // onward (skip `to`'s own entry edge — it enters
                        // the cycle from outside) plus the closing edge.
                        let mut diodes: Vec<Pos> = stack
                            .iter()
                            .skip_while(|(c, _)| c != to)
                            .skip(1)
                            .map(|(_, d)| *d)
                            .collect();
                        diodes.push(*diode);
                        diodes.sort_unstable();
                        if reported.insert(diodes.clone()) {
                            found.push(Violation::RepeaterCycle { diodes });
                        }
                    }
                    Color::White => {
                        stack.push((*to, *diode));
                        dfs(*to, edges, color, stack, found, reported);
                        stack.pop();
                    }
                    Color::Black => {}
                }
            }
        }
        color.insert(node, Color::Black);
    }

    for n in &nodes {
        if color[n] == Color::White {
            let mut stack = vec![(*n, Pos::new(0, 0, 0))];
            // The root's entry diode is a placeholder; cycles never include
            // it because cycle extraction starts at the revisited node.
            dfs(*n, &edges, &mut color, &mut stack, &mut found, &mut reported);
        }
    }
    found
}

/// Decay check: BFS through dust adjacency from every driver-adjacent dust
/// cell; dust farther than `full` (15) from all drivers cannot be powered
/// at strength. Components with no driver at all are skipped — a dead net
/// is an *open*, which static checks cannot see (only sim-vs-model or LVS
/// catches those; that is exactly why they are on the roadmap).
pub fn decay_check(cells: &BTreeMap<Pos, String>, full: u32) -> Vec<Violation> {
    // Driver dust: adjacent to a diode output, torch, lever, or sitting on
    // a strongly-powered ladder cap (block directly above a torch).
    let mut drivers: BTreeSet<Pos> = BTreeSet::new();
    for (p, b) in cells {
        if is_repeater(b) || is_comparator(b) {
            if let Some((_inp, outp)) = diode_io(*p, b) {
                if cells.get(&outp).is_some_and(|b| is_dust(b)) {
                    drivers.insert(outp);
                }
            }
        } else if blocks::is_torch(b) || blocks::is_lever(b) {
            for (dx, dz) in [(1, 0), (-1, 0), (0, 1), (0, -1)] {
                let q = p.offset(dx, 0, dz);
                if cells.get(&q).is_some_and(|b| is_dust(b)) {
                    drivers.insert(q);
                }
            }
            // Dust directly above (a torch powers the dust on top of it —
            // and a torch under a cap strongly powers the cap, whose top
            // dust is the ladder exit).
            for dy in [1, 2] {
                let q = p.offset(0, dy, 0);
                if cells.get(&q).is_some_and(|b| is_dust(b)) {
                    drivers.insert(q);
                }
            }
        }
    }
    // Multi-source BFS.
    let mut dist: BTreeMap<Pos, u32> = drivers.iter().map(|p| (*p, 0)).collect();
    let mut frontier: Vec<Pos> = drivers.iter().copied().collect();
    while !frontier.is_empty() {
        let mut next = Vec::new();
        for p in frontier {
            let d = dist[&p];
            for q in nets::neighbours(cells, p) {
                if !dist.contains_key(&q) {
                    dist.insert(q, d + 1);
                    next.push(q);
                }
            }
        }
        frontier = next;
    }
    let mut out = Vec::new();
    for (p, b) in cells {
        if !is_dust(b) {
            continue;
        }
        match dist.get(p) {
            Some(d) if *d > full => out.push(Violation::PowerStarved {
                at: *p,
                distance: *d,
            }),
            Some(_) => {}
            None => {} // no driver in the component: an open, not decay
        }
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::blocks::{repeater, DUST, LEVER_OFF, STONE};

    fn dust_line(ws: &mut Workspace, x0: i32, x1: i32, z: i32, label: &str) {
        for x in x0..=x1 {
            ws.dust(Pos::new(x, 1, z), label).unwrap();
        }
    }

    #[test]
    fn repeater_ring_latch_is_detected() {
        // The cout/cin ring, minimally: two dust runs joined by two
        // opposite-facing repeaters — aliased into one net, invisible to
        // the short checker, and it latches (bug provenance: rca_cells
        // 18/32). Repeater facing names its INPUT side.
        let mut ws = Workspace::new();
        dust_line(&mut ws, 0, 2, 0, "cout");
        dust_line(&mut ws, 0, 2, 2, "cout");
        // rep A: input z0 run (north side), output z2 run.
        ws.stone(Pos::new(0, 0, 1), "plain").unwrap();
        ws.put(Pos::new(0, 1, 1), &repeater("north", 1)).unwrap();
        // rep B: input z2 run (south side), output z0 run.
        ws.stone(Pos::new(2, 0, 1), "plain").unwrap();
        ws.put(Pos::new(2, 1, 1), &repeater("south", 1)).unwrap();
        let cycles = repeater_cycles(ws.cells());
        assert_eq!(cycles.len(), 1);
        match &cycles[0] {
            Violation::RepeaterCycle { diodes } => {
                assert_eq!(diodes, &vec![Pos::new(0, 1, 1), Pos::new(2, 1, 1)]);
            }
            v => panic!("unexpected: {v:?}"),
        }
    }

    #[test]
    fn forward_chain_is_not_a_cycle() {
        // dust -> rep -> dust -> rep -> dust, all one direction: clean.
        let mut ws = Workspace::new();
        dust_line(&mut ws, 0, 1, 0, "a");
        ws.stone(Pos::new(2, 0, 0), "plain").unwrap();
        ws.put(Pos::new(2, 1, 0), &repeater("west", 1)).unwrap();
        dust_line(&mut ws, 3, 4, 0, "a");
        ws.stone(Pos::new(5, 0, 0), "plain").unwrap();
        ws.put(Pos::new(5, 1, 0), &repeater("west", 1)).unwrap();
        dust_line(&mut ws, 6, 7, 0, "a");
        assert!(repeater_cycles(ws.cells()).is_empty());
    }

    #[test]
    fn self_loop_through_one_repeater_is_a_cycle() {
        // A repeater feeding its own input component — the unclocked
        // storage element (a deliberate latch would use exactly this).
        let mut ws = Workspace::new();
        // Ring of dust around (1,1,1)..(3,1,3) minus center, with a
        // repeater bridging a cut in the ring.
        for (x, z) in [(1, 1), (2, 1), (3, 1), (3, 2), (3, 3), (2, 3), (1, 3)] {
            ws.dust(Pos::new(x, 1, z), "loop").unwrap();
        }
        // Repeater at (1,1,2): input the (1,1,3) side (south), output (1,1,1).
        ws.stone(Pos::new(1, 0, 2), "plain").unwrap();
        ws.put(Pos::new(1, 1, 2), &repeater("south", 1)).unwrap();
        let cycles = repeater_cycles(ws.cells());
        assert_eq!(cycles.len(), 1);
    }

    #[test]
    fn decay_flags_a_17_cell_unrepeated_run() {
        let mut ws = Workspace::new();
        // Lever drives x=0; dust runs x=1..=17 (17 cells): the far cells
        // sit beyond strength 15.
        ws.stone(Pos::new(0, 0, 0), "plain").unwrap();
        ws.put(Pos::new(0, 1, 0), LEVER_OFF).unwrap();
        dust_line(&mut ws, 1, 17, 0, "n");
        let v = decay_check(ws.cells(), 15);
        assert_eq!(v.len(), 1);
        assert!(matches!(
            v[0],
            Violation::PowerStarved {
                at: Pos { x: 17, .. },
                distance: 16
            }
        ));
        // A repeater mid-run fixes it.
        let mut ws2 = Workspace::new();
        ws2.stone(Pos::new(0, 0, 0), "plain").unwrap();
        ws2.put(Pos::new(0, 1, 0), LEVER_OFF).unwrap();
        dust_line(&mut ws2, 1, 8, 0, "n");
        ws2.stone(Pos::new(9, 0, 0), "plain").unwrap();
        ws2.put(Pos::new(9, 1, 0), &repeater("west", 1)).unwrap();
        dust_line(&mut ws2, 10, 17, 0, "n");
        assert!(decay_check(ws2.cells(), 15).is_empty());
    }

    #[test]
    fn drc_aggregates_all_checks() {
        let mut ws = Workspace::new();
        // A floating dust cell labelled twice-over to also short.
        ws.put(Pos::new(0, 1, 0), DUST).unwrap(); // floating
        ws.set_label(Pos::new(0, 1, 0), "a");
        ws.stone(Pos::new(1, 0, 0), "plain").unwrap();
        ws.put(Pos::new(1, 1, 0), DUST).unwrap();
        ws.set_label(Pos::new(1, 1, 0), "b"); // shorts with a
        let vs = drc(&ws, &DrcOptions::default());
        assert!(vs.iter().any(|v| matches!(v, Violation::Short { .. })));
        assert!(vs.iter().any(|v| matches!(v, Violation::Floating { .. })));
        // Stone is not a NEEDS_FLOOR block.
        assert!(!vs
            .iter()
            .any(|v| matches!(v, Violation::Floating { at, .. } if *at == Pos::new(1, 0, 0))));
        let _ = STONE;
    }
}

File: redstone-eda/drc.py
"""Design-rule checks every builder runs: shorts, support, and diode rings.

`nets.check` proves no two signals touch and `audit.audit` proves nothing
floats.  Neither can see the third defect class, which has now cost this
project three circuits:

    a directed cycle through a repeater/comparator is a STORAGE ELEMENT.

Unless the latch was built on purpose it is a bug, and it is the worst kind of
bug this toolchain produces, because every check we had stayed green:

  * `nets.check` is blind -- every cell on the ring carries the SAME label, so
    there is no short to find;
  * `audit.audit` is blind -- the ring is perfectly supported;
  * simulation is blind -- the ring latches at 15 and the world is then
    QUIESCENT, so `run_until_quiescent` sees nothing left to do and the
    exhaustive proof just reports a wrong answer somewhere.

The three occurrences:

  1. rca_cells -- two cells' port approaches formed opposite-facing repeaters
     in a ring on an aliased net; it latched the placement transient.
  2. the same shape again on the Rust side, which is why
     `crates/nucleation-routing/src/drc.rs::repeater_cycles` exists.  This
     module is the Python port of that check -- the parity gap that let (3)
     through.
  3. mult4's `m1B2` -- `REFRESH` 5 -> 14 (commit 5076520c) re-pathed the route
     along the rail it was supposed to DRIVE.  `dust_ok` allowed it (same net
     label, and own-net reuse is normally fine), but a compiled rail is fed
     THROUGH a repeater, so the reused cells sat downstream of a diode the
     route then fed from the other end:

         rail driver -> rail cell -> route tail -> drive cell -> rail driver

     Output bit 3 stuck at 1; the multiplier scored 64/256.
     `router.downstream_rail` refuses that specific contact now, and
     `test_diode_ring` guards it -- but only for `router.py`, and only for a
     route joining a rail.  THIS check is the general rule: it does not care
     how the ring was formed or which builder formed it.

The rule, stated over geometry alone (no simulation, no netlist):

    build the conduction graph -- dust components are single nodes (union-find
    over `nets.neighbours`, so all of a net's dust collapses), each
    repeater/comparator is a DIRECTED edge from the node at its input to the
    node at its output -- then report every directed cycle.

Two conduction paths beyond bare dust-to-dust are modelled, because the router
emits them and a ring through one latches just as hard:

  * block-sandwich stations (`router.station_ok`): trunk dust -> entry block ->
    repeater -> exit block -> dust.  Dust only weak-powers the block it points
    INTO, so the input is read collinearly through the entry block; the exit
    block is STRONGLY powered and re-emits 15 to every adjacent dust, so the
    output fans out to the whole 6-neighbourhood (probe_station S_exit_*).
  * diode -> block -> diode with no dust between them, via the same
    strongly-powered exit block.

Torches are deliberately NOT diodes here: an inverter ring oscillates rather
than latching, and the NOR-latch gadgets in `rca_cells` / `seq_cells` are built
from torches on purpose.  Keeping torches out matches the Rust check and keeps
this rule free of false positives, so it can run on every build.

Cost is one pass over `cells` plus an iterative DFS over dust components --
geometry only, no simulation.  It is cheap enough to be unconditional (mult4's
20k-block design: tens of milliseconds), which is the whole point: a check that
is expensive is a check that gets skipped, and this defect class is exactly the
one that survives every check you skip.

Usage -- one line next to the audit + nets pair the builders already spell out:

    rings_ok = drc.check_rings("mult4", b.cells)
    if any(problems.values()) or shorts or not rings_ok:
        return 1

`expect` accepts a deliberate latch count for a design that really does store
state in a ring (`seq_counter` declares one per bit); it must be given
explicitly, per build, with a reason.  `drc.check` returns all three findings at
once for a caller that would rather inspect than print.
"""
import audit
import nets

DUST_KEY = nets.DUST_KEY

FACING_VEC = {"north": (0, 0, -1), "south": (0, 0, 1),
              "east": (1, 0, 0), "west": (-1, 0, 0)}


def is_dust(block):
    return block is not None and DUST_KEY in block


def is_diode(block):
    """Repeater or comparator: a one-way gate that can hold a ring high."""
    return block is not None and ("minecraft:repeater" in block
                                 or "minecraft:comparator" in block)


def facing_of(block):
    if block is None or "facing=" not in block:
        return None
    return block.split("facing=")[1].split(",")[0].rstrip("]")


def diode_io(pos, block):
    """(input cell, output cell) of the diode at `pos`.

    `facing` names the INPUT side: `repeater[facing=west]` reads -X and
    conducts toward +X (verified, rs.repeater's docstring).
    """
    f = facing_of(block)
    v = FACING_VEC.get(f)
    if v is None:
        return None
    x, y, z = pos
    dx, dy, dz = v
    return (x + dx, y + dy, z + dz), (x - dx, y - dy, z - dz)


def _components(cells):
    """Union-find over dust: pos -> component root."""
    parent = {}

    def find(a):
        parent.setdefault(a, a)
        while parent[a] != a:
            parent[a] = parent[parent[a]]
            a = parent[a]
        return a

    dust = [p for p, b in cells.items() if DUST_KEY in b]
    for p in dust:
        find(p)
        for q in nets.neighbours(cells, p):
            ra, rb = find(p), find(q)
            if ra != rb:
                parent[ra] = rb
    return find


def _adjacent(pos):
    x, y, z = pos
    return ((x + 1, y, z), (x - 1, y, z), (x, y, z + 1), (x, y, z - 1),
            (x, y + 1, z), (x, y - 1, z))


def conduction_graph(cells):
    """Directed graph whose cycles are latches.

    Nodes are ``("d", root)`` for a dust component and ``("b", pos)`` for a
    solid block that some diode STRONGLY powers.  Edges out of a diode are
    labelled with that diode's position; edges out of a block node are plain
    conduction (strong power re-emitted to adjacent dust).  A block node's only
    incoming edges come from diodes, and dust-to-dust conduction is collapsed
    into the component, so every cycle in this graph contains at least one
    diode -- which is exactly the thing that makes it latch.

    Returns ``{node: [(node, diode_pos_or_None), ...]}``.
    """
    find = _components(cells)
    diodes = [(p, b) for p, b in cells.items() if is_diode(b)]

    # Blocks a diode strongly powers become nodes: they re-emit 15.
    strong = set()
    for p, b in diodes:
        io = diode_io(p, b)
        if io is None:
            continue
        out = io[1]
        if not is_dust(cells.get(out)) and nets.is_solid(cells, out):
            strong.add(out)

    edges = {}

    def add(src, dst, diode):
        # self-loops are kept on purpose: src == dst is the mult4 shape, a
        # diode whose input and output landed in the SAME dust component
        edges.setdefault(src, []).append((dst, diode))

    def node_at(pos, collinear_probe=None):
        """The conduction node a diode reads/drives at `pos`."""
        if is_dust(cells.get(pos)):
            return ("d", find(pos))
        if pos in strong:
            return ("b", pos)
        if collinear_probe is not None and nets.is_solid(cells, pos):
            # station entry block: only the dust POINTING INTO it powers it
            if is_dust(cells.get(collinear_probe)):
                return ("d", find(collinear_probe))
        return None

    for p, b in diodes:
        io = diode_io(p, b)
        if io is None:
            continue
        inp, out = io
        v = FACING_VEC[facing_of(b)]
        beyond = (inp[0] + v[0], inp[1] + v[1], inp[2] + v[2])
        src = node_at(inp, collinear_probe=beyond)
        dst = node_at(out)
        if src is None or dst is None:
            continue
        add(src, dst, p)

    # Strongly-powered blocks re-emit to every adjacent dust cell.
    for blk in strong:
        for q in _adjacent(blk):
            if is_dust(cells.get(q)):
                add(("b", blk), ("d", find(q)), None)

    return edges


def repeater_cycles(cells):
    """Every directed cycle through a repeater/comparator.

    Returns a list of tuples of diode positions, sorted and de-duplicated, so
    the result is deterministic and a caller can print it verbatim.
    """
    edges = conduction_graph(cells)
    nodes = set(edges)
    for succs in edges.values():
        for to, _d in succs:
            nodes.add(to)

    WHITE, GREY, BLACK = 0, 1, 2
    color = {n: WHITE for n in nodes}
    found, reported = [], set()

    for root in sorted(nodes):
        if color[root] != WHITE:
            continue
        color[root] = GREY
        path = [root]
        entry = [None]                      # diode each path node was entered by
        stack = [iter(edges.get(root, ()))]
        while stack:
            nxt = next(stack[-1], None)
            if nxt is None:
                color[path[-1]] = BLACK
                path.pop()
                entry.pop()
                stack.pop()
                continue
            to, diode = nxt
            c = color.get(to, WHITE)
            if c == GREY:
                # the cycle runs from `to` forward along the path and closes
                # through this edge; `to`'s own entry edge came from outside
                i = path.index(to)
                ring = [d for d in entry[i + 1:] if d is not None]
                if diode is not None:
                    ring.append(diode)
                key = tuple(sorted(ring))
                if key and key not in reported:
                    reported.add(key)
                    found.append(key)
            elif c == WHITE:
                color[to] = GREY
                path.append(to)
                entry.append(diode)
                stack.append(iter(edges.get(to, ())))
    return found


def check_rings(name, cells, expect=0, quiet=False):
    """The ring rule alone, for a builder that already prints its own
    audit/nets report.  Returns True when the geometry holds no latch.

    `expect` is a deliberate latch count -- a design that really does store
    state states it here, with a reason, at the call site.  Never raise it to
    silence a surprise: an unexplained ring means the build is broken in the
    one way nothing else can see.
    """
    cyc = repeater_cycles(cells)
    if len(cyc) == expect:
        if not quiet:
            print("diode rings: %d%s" % (len(cyc), " (declared)" if expect else ""))
        return True
    if not quiet:
        print("DIODE RING x%d (expected %d) in %s -- a ring holding a repeater "
              "LATCHES at 15: the net reads high forever, no input can clear "
              "it, and the world is quiescent so simulation cannot see it"
              % (len(cyc), expect, name))
        for ring in cyc[:6]:
            print("   diodes:", ", ".join(str(p) for p in ring))
    return False


def check(cells, labels, aliases=()):
    """All three static rules at once, for a caller that wants the findings
    rather than a printed report: shorts, support, rings."""
    return {
        "problems": audit.audit(cells),
        "shorts": nets.check(cells, labels, aliases),
        "cycles": repeater_cycles(cells),
    }


if __name__ == "__main__":
    # Ring report for the builds that are cheap to construct.  The real proof
    # lives in test_diode_ring.py, which also rebuilds the PRE-FIX mult4 tree
    # and confirms the rule fires there.
    import time

    def report(label, cells):
        t0 = time.time()
        rings = repeater_cycles(cells)
        print("%-14s %7d cells  %5.0f ms  rings=%d%s"
              % (label, len(cells), 1000 * (time.time() - t0), len(rings),
                 "" if not rings else "  %s" % (rings[:2],)))

    import build_adder as ad
    report("adder4", ad.build(4)[0].cells)
    import mult4
    report("mult4", mult4.build()["b"].cells)
